plugins/amazonq/shared/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/lsp/AmazonQLspService.kt

```diff
@@ -195,6 +235,34 @@ class AmazonQLspService(private val project: Project, private val cs: CoroutineS
         instance = start()
     }
 
+    private suspend fun waitForRestartSlot(): Unit = restartMutex.withLock {
+        val currentTime = System.currentTimeMillis()
+
+        while (restartTimestamps.isNotEmpty() &&
+            currentTime - restartTimestamps.first() > RESTART_WINDOW_MS
+        ) {
+            restartTimestamps.removeFirst()
+        }
+
+        if (restartTimestamps.size < MAX_RESTARTS) {
+            restartTimestamps.addLast(currentTime)
+            return
+        }
+
+        val oldestTimestamp = restartTimestamps.first()
+        val waitTimeMs = (oldestTimestamp + RESTART_WINDOW_MS) - currentTime
+
+        LOG.info { "Rate limit reached for LSP server restarts. Waiting ${waitTimeMs}ms for next available slot." }
+
+        restartMutex.unlock()
+        delay(waitTimeMs.milliseconds)
+
+        // After waiting, recursively call this function to check again
+        // (in case conditions changed while we were waiting)
+        waitForRestartSlot()
+        restartMutex.lock()
```

Comment:
seems redundant since the lock will be released right after this

Reply:
yea, I will need to change withLock code then.